# Angela Shane

## SQL Homework 1

### Database Setup
The following queries run against the [Pagila database](https://www.postgresql.org/ftp/projects/pgFoundry/dbsamples/pagila/pagila/).  Run the following steps to setup the database
1. Download pagila-0.10.1.zip
2. Unzip the zip file
3. From the command line:
  * change directories to the unzipped files
  * run psql
4. run each sql file by doing: \i filename.sql
  * Run files in this order: pagila-schema.sql, pagila-insert-data.sql, pagila-data.sql


In [ ]:
import pandas as pd
import psycopg2
import 

In [97]:
# parameters
host="localhost"
database="pagila"
user="postgres"
password="************"

# Creating a connection
conn_str = "host={} dbname={} user={} password={}".format(host, database, user, password)
conn = psycopg2.connect(conn_str)

### 1a. You need a list of all the actors’ first name and last name

In [98]:
df = pd.read_sql('SELECT first_name, last_name FROM actor', con=conn)
df.head()

first_name     last_name
0   PENELOPE       GUINESS
1       NICK      WAHLBERG
2         ED         CHASE
3   JENNIFER         DAVIS
4     JOHNNY  LOLLOBRIGIDA

### 1b. Display the first and last name of each actor in a single column in upper case letters. Name the column Actor Name

In [99]:
df = pd.read_sql('SELECT UPPER(first_name || \' \' || last_name) as "Actor Name" FROM actor', con=conn)
df.head()

Actor Name
0     PENELOPE GUINESS
1        NICK WAHLBERG
2             ED CHASE
3       JENNIFER DAVIS
4  JOHNNY LOLLOBRIGIDA

### 2a. You need to find the id, first name, and last name of an actor, of whom you know only the first name of "Joe." What is one query would you use to obtain this information?

In [100]:
df = pd.read_sql('SELECT actor_id, first_name, last_name FROM actor WHERE first_name ilike \'Joe\'', con=conn)
df

actor_id first_name last_name
0         9        JOE     SWANK

### 2b. Find all actors whose last name contain the letters GEN. Make this case insensitive


In [101]:
df = pd.read_sql('SELECT actor_id, first_name, last_name FROM actor WHERE last_name ilike \'%GEN%\'', con=conn)
df

actor_id first_name  last_name
0        14     VIVIEN     BERGEN
1        41      JODIE  DEGENERES
2       107       GINA  DEGENERES
3       166       NICK  DEGENERES

### 2c. Find all actors whose last names contain the letters LI. This time, order the rows by last name and first name, in that order. Make this case insensitive.


In [102]:
df = pd.read_sql('SELECT actor_id, first_name, last_name FROM actor WHERE last_name ilike \'%LI%\' ORDER BY last_name, first_name', con=conn)
df

actor_id first_name last_name
0        86       GREG   CHAPLIN
1        82      WOODY     JOLIE
2        34     AUDREY   OLIVIER
3        15       CUBA   OLIVIER
4       172    GROUCHO  WILLIAMS
5       137     MORGAN  WILLIAMS
6        72       SEAN  WILLIAMS
7        83        BEN    WILLIS
8        96       GENE    WILLIS
9       164   HUMPHREY    WILLIS

### 2d. Using IN, display the country_id and country columns of the following countries: Afghanistan, Bangladesh, and China:


In [103]:
df = pd.read_sql('SELECT * FROM country WHERE country IN (\'Afghanistan\',\'Bangladesh\', \'China\')', con=conn)
df

country_id      country         last_update
0           1  Afghanistan 2006-02-15 09:44:00
1          12   Bangladesh 2006-02-15 09:44:00
2          23        China 2006-02-15 09:44:00

### 3a. Add a middle_name column to the table actor. Specify the appropriate column type


In [104]:
cur = conn.cursor()
cur.execute('ALTER TABLE actor ADD COLUMN middle_name varchar(45);')
df = pd.read_sql('SELECT * FROM actor', con=conn)
df.head()

actor_id first_name     last_name         last_update middle_name
0         1   PENELOPE       GUINESS 2006-02-15 09:34:33        None
1         2       NICK      WAHLBERG 2006-02-15 09:34:33        None
2         3         ED         CHASE 2006-02-15 09:34:33        None
3         4   JENNIFER         DAVIS 2006-02-15 09:34:33        None
4         5     JOHNNY  LOLLOBRIGIDA 2006-02-15 09:34:33        None

### 3b. You realize that some of these actors have tremendously long last names. Change the data type of the middle_name column to something that can hold more than varchar.


In [105]:
cur = conn.cursor()
cur.execute('ALTER TABLE actor ALTER COLUMN middle_name TYPE text;')
df = pd.read_sql('SELECT * FROM actor', con=conn)
df.head()

actor_id first_name     last_name         last_update middle_name
0         1   PENELOPE       GUINESS 2006-02-15 09:34:33        None
1         2       NICK      WAHLBERG 2006-02-15 09:34:33        None
2         3         ED         CHASE 2006-02-15 09:34:33        None
3         4   JENNIFER         DAVIS 2006-02-15 09:34:33        None
4         5     JOHNNY  LOLLOBRIGIDA 2006-02-15 09:34:33        None

### 3c. Now write a query that would remove the middle_name column.


In [106]:
cur = conn.cursor()
cur.execute('ALTER TABLE actor DROP COLUMN IF EXISTS middle_name;')
df = pd.read_sql('SELECT * FROM actor', con=conn)
df.head()

actor_id first_name     last_name         last_update
0         1   PENELOPE       GUINESS 2006-02-15 09:34:33
1         2       NICK      WAHLBERG 2006-02-15 09:34:33
2         3         ED         CHASE 2006-02-15 09:34:33
3         4   JENNIFER         DAVIS 2006-02-15 09:34:33
4         5     JOHNNY  LOLLOBRIGIDA 2006-02-15 09:34:33

### 4a. List the last names of actors, as well as how many actors have that last name.


In [108]:
df = pd.read_sql('SELECT last_name, COUNT(*) FROM actor GROUP BY last_name ORDER BY last_name', con=conn)
df

last_name  count
0         AKROYD      3
1          ALLEN      3
2        ASTAIRE      1
3         BACALL      1
4         BAILEY      2
5           BALE      1
6           BALL      1
7      BARRYMORE      1
8       BASINGER      1
9         BENING      2
10        BERGEN      1
11       BERGMAN      1
12         BERRY      3
13         BIRCH      1
14         BLOOM      1
15        BOLGER      2
16       BRIDGES      1
17         BRODY      2
18       BULLOCK      1
19          CAGE      2
20        CARREY      1
21       CHAPLIN      1
22         CHASE      2
23         CLOSE      1
24       COSTNER      1
25      CRAWFORD      2
26        CRONYN      2
27         CROWE      1
28        CRUISE      1
29          CRUZ      1
..           ...    ...
91         POSEY      1
92       PRESLEY      1
93      REYNOLDS      1
94         RYDER      1
95   SILVERSTONE      2
96       SINATRA      1
97      SOBIESKI      1
98      STALLONE      1
99        STREEP      2
100       SUVARI      1
101        SWANK      1
102        TANDY      2
103       TAUTOU      1
104       TEMPLE      4
105        TOMEI      1
106         TORN      3
107        TRACY      2
108       VOIGHT      1
109     WAHLBERG      2
110       WALKEN      1
111        WAYNE      1
112         WEST      2
113     WILLIAMS      3
114       WILLIS      3
115       WILSON      1
116      WINSLET      2
117  WITHERSPOON      1
118         WOOD      2
119         WRAY      1
120    ZELLWEGER      3

[121 rows x 2 columns]

### 4b. List last names of actors and the number of actors who have that last name, but only for names that are shared by at least two actors


In [109]:
df = pd.read_sql('SELECT last_name, COUNT(*) FROM actor GROUP BY last_name HAVING COUNT(*) >= 2 ORDER BY last_name', con=conn)
df

last_name  count
0        AKROYD      3
1         ALLEN      3
2        BAILEY      2
3        BENING      2
4         BERRY      3
5        BOLGER      2
6         BRODY      2
7          CAGE      2
8         CHASE      2
9      CRAWFORD      2
10       CRONYN      2
11        DAVIS      3
12         DEAN      2
13          DEE      2
14    DEGENERES      3
15        DENCH      2
16         DEPP      2
17      DUKAKIS      2
18      FAWCETT      2
19      GARLAND      3
20      GOODING      2
21      GUINESS      3
22      HACKMAN      2
23       HARRIS      3
24      HOFFMAN      3
25      HOPKINS      3
26       HOPPER      2
27      JACKMAN      2
28    JOHANSSON      3
29       KEITEL      3
30       KILMER      5
31  MCCONAUGHEY      2
32     MCKELLEN      2
33      MCQUEEN      2
34       MONROE      2
35       MOSTEL      2
36       NEESON      2
37        NOLTE      4
38      OLIVIER      2
39      PALTROW      2
40         PECK      3
41         PENN      2
42  SILVERSTONE      2
43       STREEP      2
44        TANDY      2
45       TEMPLE      4
46         TORN      3
47        TRACY      2
48     WAHLBERG      2
49         WEST      2
50     WILLIAMS      3
51       WILLIS      3
52      WINSLET      2
53         WOOD      2
54    ZELLWEGER      3

### 4c. Oh, no! The actor HARPO WILLIAMS was accidentally entered in the actor table as GROUCHO WILLIAMS. Write a query to fix the record.


In [110]:
df = pd.read_sql('SELECT * FROM actor WHERE first_name ilike \'Groucho\' and last_name ilike \'Williams\'', con=conn)
df.head()

actor_id first_name last_name         last_update
0       172    GROUCHO  WILLIAMS 2006-02-15 09:34:33

In [111]:
cur = conn.cursor()
cur.execute('UPDATE actor SET first_name = \'HARPO\' WHERE first_name ilike \'Groucho\' and last_name ilike \'Williams\';')


In [112]:
df = pd.read_sql('SELECT * FROM actor WHERE actor_id = 172', con=conn)
df

actor_id first_name last_name                last_update
0       172      HARPO  WILLIAMS 2017-09-24 17:30:14.644315

### 4d. Perhaps we were too hasty in changing GROUCHO to HARPO. It turns out that GROUCHO was the correct name after all! 

In [121]:
cur = conn.cursor()
cur.execute('UPDATE actor SET first_name = CASE WHEN first_name = \'HARPO\' THEN \'GROUCHO\' ELSE \'MUCHO GROUCHO\' END WHERE actor_id = 172;')
#cur.execute('UPDATE actor SET first_name = \'HARPO\' WHERE actor_id = 172;')


In [122]:
df = pd.read_sql('SELECT * FROM actor WHERE actor_id = 172', con=conn)
df

actor_id first_name last_name                last_update
0       172    GROUCHO  WILLIAMS 2017-09-24 17:30:14.644315

### 5a. What’s the difference between a left join and a right join?


#### Left Join
A left join returns all rows in the left table and matching rows in the right table.
In the example below, the address table has 603 rows.
In a left join from the address table to the city table, each address row is returned with the matching city value from the city table.   The left join returns 603 rows.


In [131]:
df = pd.read_sql('SELECT COUNT(*) FROM address;', con=conn)
df

count
0    603

In [132]:
df = pd.read_sql('SELECT a.address_id, a.city_id, c.city FROM address a LEFT JOIN city c ON a.city_id = c.city_id ORDER BY a.address_id DESC;', con=conn)
df

address_id  city_id               city
0           605      537              Tieli
1           604      296           Lausanne
2           603      503            Sullana
3           602      401             Patras
4           601      242            Jinzhou
5           600      241             Jining
6           599      177       Garden Grove
7           598      512        Szkesfehrvr
8           597      248       Juiz de Fora
9           596       14          al-Manama
10          595      311               Loja
11          594      574            Weifang
12          593      106             Celaya
13          592      546            Tsaotun
14          591      424            Quilmes
15          590       54             Banjul
16          589      494        Southampton
17          588      212  Huejutla de Reyes
18          587      246         Jos Azueta
19          586      219             Iligan
20          585      344              Mosul
21          584      103             Carmen
22          583      526            Tanshui
23          582      168           Fengshan
24          581      131         Cuernavaca
25          580       10           Akishima
26          579      598        Zhezqazghan
27          578      133        Czestochowa
28          577      409              Plock
29          576      341           Monclova
..          ...      ...                ...
573          30      121  Citt del Vaticano
574          29      472          Shikarpur
575          28      128             Crdoba
576          27      303            Liepaja
577          26      443                Sal
578          25      525            Tangail
579          24      327             Mardan
580          23      267          Kimberley
581          22      252             Kaduna
582          21      156             Elista
583          20      495    Southend-on-Sea
584          19       76             Bhopal
585          18      120     Citrus Heights
586          17      384           Osmaniye
587          16      582       Yamuna Nagar
588          15      440         Sagamihara
589          14      162            Esfahan
590          13      329             Masqat
591          12      200           Hamilton
592          11      280         Kragujevac
593          10      295             Laredo
594           9      361             Nantou
595           8      349           Myingyan
596           7       38            Athenai
597           6      449     San Bernardino
598           5      463             Sasebo
599           4      576          Woodridge
600           3      300         Lethbridge
601           2      576          Woodridge
602           1      300         Lethbridge

[603 rows x 3 columns]

#### Right Joint
A right join returns all rows in the right table and matching rows in the left table. In the example below, the city table has 600 rows.
In a right join to the address table, each city row is returned with the matching address value from the address table.   The right join returns 604 rows.  The city London has no address with a city in London.


In [133]:
df = pd.read_sql('SELECT COUNT(*) FROM city;', con=conn)
df

count
0    600

In [134]:
df = pd.read_sql('SELECT a.address_id, a.city_id, c.city FROM address a RIGHT JOIN city c ON a.city_id = c.city_id ORDER BY a.address_id DESC;', con=conn)
df

address_id  city_id               city
0           NaN      NaN             London
1         605.0    537.0              Tieli
2         604.0    296.0           Lausanne
3         603.0    503.0            Sullana
4         602.0    401.0             Patras
5         601.0    242.0            Jinzhou
6         600.0    241.0             Jining
7         599.0    177.0       Garden Grove
8         598.0    512.0        Szkesfehrvr
9         597.0    248.0       Juiz de Fora
10        596.0     14.0          al-Manama
11        595.0    311.0               Loja
12        594.0    574.0            Weifang
13        593.0    106.0             Celaya
14        592.0    546.0            Tsaotun
15        591.0    424.0            Quilmes
16        590.0     54.0             Banjul
17        589.0    494.0        Southampton
18        588.0    212.0  Huejutla de Reyes
19        587.0    246.0         Jos Azueta
20        586.0    219.0             Iligan
21        585.0    344.0              Mosul
22        584.0    103.0             Carmen
23        583.0    526.0            Tanshui
24        582.0    168.0           Fengshan
25        581.0    131.0         Cuernavaca
26        580.0     10.0           Akishima
27        579.0    598.0        Zhezqazghan
28        578.0    133.0        Czestochowa
29        577.0    409.0              Plock
..          ...      ...                ...
574        30.0    121.0  Citt del Vaticano
575        29.0    472.0          Shikarpur
576        28.0    128.0             Crdoba
577        27.0    303.0            Liepaja
578        26.0    443.0                Sal
579        25.0    525.0            Tangail
580        24.0    327.0             Mardan
581        23.0    267.0          Kimberley
582        22.0    252.0             Kaduna
583        21.0    156.0             Elista
584        20.0    495.0    Southend-on-Sea
585        19.0     76.0             Bhopal
586        18.0    120.0     Citrus Heights
587        17.0    384.0           Osmaniye
588        16.0    582.0       Yamuna Nagar
589        15.0    440.0         Sagamihara
590        14.0    162.0            Esfahan
591        13.0    329.0             Masqat
592        12.0    200.0           Hamilton
593        11.0    280.0         Kragujevac
594        10.0    295.0             Laredo
595         9.0    361.0             Nantou
596         8.0    349.0           Myingyan
597         7.0     38.0            Athenai
598         6.0    449.0     San Bernardino
599         5.0    463.0             Sasebo
600         4.0    576.0          Woodridge
601         3.0    300.0         Lethbridge
602         2.0    576.0          Woodridge
603         1.0    300.0         Lethbridge

[604 rows x 3 columns]

### 5b. What about an inner join and an outer join?

### Inner Join
An inner join returns only matching rows between the left and right table.

In [136]:
df = pd.read_sql('SELECT a.address_id, a.city_id, c.city FROM address a JOIN city c ON a.city_id = c.city_id ORDER BY a.address_id DESC;', con=conn)
df

address_id  city_id               city
0           605      537              Tieli
1           604      296           Lausanne
2           603      503            Sullana
3           602      401             Patras
4           601      242            Jinzhou
5           600      241             Jining
6           599      177       Garden Grove
7           598      512        Szkesfehrvr
8           597      248       Juiz de Fora
9           596       14          al-Manama
10          595      311               Loja
11          594      574            Weifang
12          593      106             Celaya
13          592      546            Tsaotun
14          591      424            Quilmes
15          590       54             Banjul
16          589      494        Southampton
17          588      212  Huejutla de Reyes
18          587      246         Jos Azueta
19          586      219             Iligan
20          585      344              Mosul
21          584      103             Carmen
22          583      526            Tanshui
23          582      168           Fengshan
24          581      131         Cuernavaca
25          580       10           Akishima
26          579      598        Zhezqazghan
27          578      133        Czestochowa
28          577      409              Plock
29          576      341           Monclova
..          ...      ...                ...
573          30      121  Citt del Vaticano
574          29      472          Shikarpur
575          28      128             Crdoba
576          27      303            Liepaja
577          26      443                Sal
578          25      525            Tangail
579          24      327             Mardan
580          23      267          Kimberley
581          22      252             Kaduna
582          21      156             Elista
583          20      495    Southend-on-Sea
584          19       76             Bhopal
585          18      120     Citrus Heights
586          17      384           Osmaniye
587          16      582       Yamuna Nagar
588          15      440         Sagamihara
589          14      162            Esfahan
590          13      329             Masqat
591          12      200           Hamilton
592          11      280         Kragujevac
593          10      295             Laredo
594           9      361             Nantou
595           8      349           Myingyan
596           7       38            Athenai
597           6      449     San Bernardino
598           5      463             Sasebo
599           4      576          Woodridge
600           3      300         Lethbridge
601           2      576          Woodridge
602           1      300         Lethbridge

[603 rows x 3 columns]

### Full Outer Join
A full outer join will return matching rows from the left and right table and non-matching rows returning null for non-matching columns.  In the example below, London is in the city table but has no matching address value and returns null.

In [140]:
df = pd.read_sql('SELECT a.address_id, a.city_id, c.city FROM address a FULL OUTER JOIN city c ON a.city_id = c.city_id ORDER BY a.address_id DESC;', con=conn)
df

address_id  city_id               city
0           NaN      NaN             London
1         605.0    537.0              Tieli
2         604.0    296.0           Lausanne
3         603.0    503.0            Sullana
4         602.0    401.0             Patras
5         601.0    242.0            Jinzhou
6         600.0    241.0             Jining
7         599.0    177.0       Garden Grove
8         598.0    512.0        Szkesfehrvr
9         597.0    248.0       Juiz de Fora
10        596.0     14.0          al-Manama
11        595.0    311.0               Loja
12        594.0    574.0            Weifang
13        593.0    106.0             Celaya
14        592.0    546.0            Tsaotun
15        591.0    424.0            Quilmes
16        590.0     54.0             Banjul
17        589.0    494.0        Southampton
18        588.0    212.0  Huejutla de Reyes
19        587.0    246.0         Jos Azueta
20        586.0    219.0             Iligan
21        585.0    344.0              Mosul
22        584.0    103.0             Carmen
23        583.0    526.0            Tanshui
24        582.0    168.0           Fengshan
25        581.0    131.0         Cuernavaca
26        580.0     10.0           Akishima
27        579.0    598.0        Zhezqazghan
28        578.0    133.0        Czestochowa
29        577.0    409.0              Plock
..          ...      ...                ...
574        30.0    121.0  Citt del Vaticano
575        29.0    472.0          Shikarpur
576        28.0    128.0             Crdoba
577        27.0    303.0            Liepaja
578        26.0    443.0                Sal
579        25.0    525.0            Tangail
580        24.0    327.0             Mardan
581        23.0    267.0          Kimberley
582        22.0    252.0             Kaduna
583        21.0    156.0             Elista
584        20.0    495.0    Southend-on-Sea
585        19.0     76.0             Bhopal
586        18.0    120.0     Citrus Heights
587        17.0    384.0           Osmaniye
588        16.0    582.0       Yamuna Nagar
589        15.0    440.0         Sagamihara
590        14.0    162.0            Esfahan
591        13.0    329.0             Masqat
592        12.0    200.0           Hamilton
593        11.0    280.0         Kragujevac
594        10.0    295.0             Laredo
595         9.0    361.0             Nantou
596         8.0    349.0           Myingyan
597         7.0     38.0            Athenai
598         6.0    449.0     San Bernardino
599         5.0    463.0             Sasebo
600         4.0    576.0          Woodridge
601         3.0    300.0         Lethbridge
602         2.0    576.0          Woodridge
603         1.0    300.0         Lethbridge

[604 rows x 3 columns]

### 5c. When would you use rank? 
Rank assigns a unique number within your ordered partition. Ties are assigned the same rank, with the next ranking(s) skipped. So, if you have 3 items at rank 2, the next rank listed would be ranked 5.

### 5d. What about dense_rank? 
DENSE_RANK assigns a unique number within your ordered partition, but the ranks are consecutive. No ranks are skipped if there are ranks with multiple items.

### 5e. When would you use a subquery in a select? 
A subqueries are a tool for performing operations in multiple steps.  For example this query is done in two separate queries:

In [142]:
df = pd.read_sql('SELECT AVG(p.amount) FROM payment p;', con=conn)
df

avg
0  4.200667

In [143]:
df = pd.read_sql('SELECT p.customer_id, p.amount FROM payment p WHERE p.amount > 4.2006673312979002;', con=conn)
df


customer_id  amount
0             269    6.99
1             269    4.99
2             270    4.99
3             271    8.99
4             272    6.99
5             272    6.99
6             274    4.99
7             274    5.99
8             274    4.99
9             276   10.99
10            277    6.99
11            278    4.99
12            280    4.99
13            282    6.99
14            284    6.99
15            286    6.99
16            288    6.99
17            288    4.99
18            288    5.99
19            291    4.99
20            291    4.99
21            293    4.99
22            293    8.99
23            296    4.99
24            296    5.99
25            296    4.99
26            299    5.99
27            299    8.99
28            300    4.99
29            301    4.99
...           ...     ...
7716           22    4.99
7717           42    5.98
7718           44    4.99
7719           52    4.99
7720           53    7.98
7721           56    4.99
7722           60    9.98
7723           64    4.99
7724           75    8.97
7725           83    4.99
7726           87    4.99
7727           91    4.99
7728           94    4.99
7729          108    4.99
7730          114    4.99
7731          120    4.99
7732          152    4.99
7733          155    7.98
7734          163    7.98
7735          178    4.99
7736          180    4.99
7737          190    4.99
7738          192    4.99
7739          208    5.98
7740          211    4.99
7741          216    5.98
7742          219    4.99
7743          227    4.99
7744          244    4.99
7745          252    4.99

[7746 rows x 2 columns]

A subquery can be used to perform the same queries in a single query.

In [144]:
df = pd.read_sql('SELECT p.customer_id, p.amount FROM payment p WHERE p.amount > (SELECT AVG(p.amount) FROM payment p);', con=conn)
df


customer_id  amount
0             269    6.99
1             269    4.99
2             270    4.99
3             271    8.99
4             272    6.99
5             272    6.99
6             274    4.99
7             274    5.99
8             274    4.99
9             276   10.99
10            277    6.99
11            278    4.99
12            280    4.99
13            282    6.99
14            284    6.99
15            286    6.99
16            288    6.99
17            288    4.99
18            288    5.99
19            291    4.99
20            291    4.99
21            293    4.99
22            293    8.99
23            296    4.99
24            296    5.99
25            296    4.99
26            299    5.99
27            299    8.99
28            300    4.99
29            301    4.99
...           ...     ...
7716           22    4.99
7717           42    5.98
7718           44    4.99
7719           52    4.99
7720           53    7.98
7721           56    4.99
7722           60    9.98
7723           64    4.99
7724           75    8.97
7725           83    4.99
7726           87    4.99
7727           91    4.99
7728           94    4.99
7729          108    4.99
7730          114    4.99
7731          120    4.99
7732          152    4.99
7733          155    7.98
7734          163    7.98
7735          178    4.99
7736          180    4.99
7737          190    4.99
7738          192    4.99
7739          208    5.98
7740          211    4.99
7741          216    5.98
7742          219    4.99
7743          227    4.99
7744          244    4.99
7745          252    4.99

[7746 rows x 2 columns]

### 5f. When would you use a right join?
When you want all rows from the right table and matching values from the left table.

### 5g. When would you use an inner join over an outer join?
When you only want rows that match in the left and right table use an inner join.


### 5h. What’s the difference between a left outer and a left join
A left outer and a left join are the same thing. Both return the rows in the left table and matching values in the right table. The word Outer is optional.

### 5i. When would you use a group by?
A group by is useful anytime you want to perform an aggregate calculation like max, min, sum, count... For example, if you wanted to know how many addresses are in each city, you can use the count function and group by city.


In [148]:
df = pd.read_sql('SELECT city_id, count(*) as address_cnt FROM address GROUP BY city_id ORDER BY address_cnt DESC;', con=conn)
df.head()

city_id  address_cnt
0       42            2
1      312            2
2      300            2
3      576            2
4      264            1

### 5j. Describe how you would do data reformatting
You can use string functions like replace, position and substring to manipulate string values.  For example, you could replace the word "Drive" with "Dr." in the address.

In [153]:
query = ("SELECT address,\n"
             "OVERLAY(address PLACING \'Dr. \' FROM POSITION(\'Drive\' IN address) FOR 5)\n"
         "FROM address\n"
         "WHERE address LIKE \'%Drive%\';\n")
df = pd.read_sql(query, con=conn)
df.head()

address              overlay
0     47 MySakila Drive     47 MySakila Dr. 
1  1411 Lillydale Drive  1411 Lillydale Dr. 
2     613 Korolev Drive     613 Korolev Dr. 
3        1531 Sal Drive        1531 Sal Dr. 
4      391 Callao Drive      391 Callao Dr.

### 5k. When would you use a with clause?
A with clause is like an in-line view, that can only be executed once.  A with clause allows you to make a complicate query easier to read and also allows you to chain selects.  Below is an example with clause:

In [154]:
query = ("WITH s AS(\n"
            "SELECT *\n"
            "FROM staff\n"
        ")\n"
        "SELECT s.staff_id, st.store_id\n"
        "FROM s\n"
        "LEFT JOIN store st ON st.store_id = s.store_id;\n")

df = pd.read_sql(query, con=conn)
df.head()



staff_id  store_id
0         1         1
1         2         2

### 5l. bonus: When would you use a self join?
You can use a self join to join a table to itself.  For example, you can use a self join to delete duplicates in a table. The example below shows duplicate actors with the same last name.

In [188]:
query = ("SELECT DISTINCT a.actor_id, a.last_name\n"
        "FROM actor a, actor b\n"
        "WHERE a.actor_id > b.actor_id\n"
            "AND a.last_name = b.last_name\n"
        "ORDER BY a.last_name;\n")
df = pd.read_sql(query, con=conn)
df

actor_id    last_name
0         92       AKROYD
1        182       AKROYD
2        145        ALLEN
3        194        ALLEN
4        190       BAILEY
5        174       BENING
6         60        BERRY
7         91        BERRY
8        185       BOLGER
9        159        BRODY
10        40         CAGE
11       176        CHASE
12       129     CRAWFORD
13       104       CRONYN
14       101        DAVIS
15       110        DAVIS
16       143         DEAN
17       148          DEE
18       107    DEGENERES
19       166    DEGENERES
20       123        DENCH
21       160         DEPP
22       188      DUKAKIS
23       199      FAWCETT
24       165      GARLAND
25       184      GARLAND
26       191      GOODING
27        90      GUINESS
28       179      GUINESS
29       175      HACKMAN
..       ...          ...
49       178       MONROE
50        99       MOSTEL
51        62       NEESON
52       122        NOLTE
53       125        NOLTE
54       150        NOLTE
55        34      OLIVIER
56        69      PALTROW
57        33         PECK
58        87         PECK
59       133         PENN
60       195  SILVERSTONE
61       116       STREEP
62       155        TANDY
63       149       TEMPLE
64       193       TEMPLE
65       200       TEMPLE
66        94         TORN
67       102         TORN
68       117        TRACY
69        95     WAHLBERG
70       197         WEST
71       137     WILLIAMS
72       172     WILLIAMS
73        96       WILLIS
74       164       WILLIS
75       147      WINSLET
76       156         WOOD
77       111    ZELLWEGER
78       186    ZELLWEGER

[79 rows x 2 columns]

### 6a. Use a JOIN to display the first and last names, as well as the address, of each staff member. Use the tables staff and address:


In [155]:
query = ("SELECT s.first_name, s.last_name, a.address\n"
        "FROM staff s\n"
        "JOIN address a on s.address_id = a.address_id;\n")
df = pd.read_sql(query, con=conn)
df.head()


first_name last_name               address
0       Mike   Hillyer     23 Workhaven Lane
1        Jon  Stephens  1411 Lillydale Drive

### 6b. Use a JOIN to display the total amount rung up by each staff member in January of 2007. Use tables staff and payment.

In [157]:
query = ("WITH x AS (\n"
            "SELECT s.first_name, s.last_name, s.staff_id, p.amount, p.payment_date\n"
            "FROM staff s\n"
            "JOIN payment p ON s.staff_id = p.staff_id\n"
            "WHERE EXTRACT(YEAR FROM p.payment_date) = 2007 AND EXTRACT(MONTH FROM p.payment_date) = 1\n"
        ")\n"
        "SELECT x.first_name, x.last_name, SUM(x.amount) as TotalPayment\n"
        "FROM x\n"
        "GROUP BY x.staff_id, x.first_name, x.last_name;\n")
df = pd.read_sql(query, con=conn)
df.head()

first_name last_name  totalpayment
0       Mike   Hillyer       2621.83
1        Jon  Stephens       2202.60

### 6c. List each film and the number of actors who are listed for that film. Use tables film_actor and film. Use inner join.

In [160]:
query = ("SELECT f.film_id, f.title, COUNT(fa.actor_id)\n"
        "FROM film f\n"
        "JOIN film_actor fa ON f.film_id = fa.film_id\n"
        "GROUP BY f.film_id;\n")
df = pd.read_sql(query, con=conn)
df.head(10)

film_id                  title  count
0      251    DRAGONFLY STRANGERS      3
1      106  BULWORTH COMMANDMENTS      4
2      285       ENGLISH BULWORTH      5
3      120      CARIBBEAN LIBERTY      8
4      681        PIRATES ROXANNE      1
5      866           SUNSET RACER      3
6      887          THIEF PELICAN      7
7      264           DWARFS ALTER      1
8      802            SKY MIRACLE     12
9      664          PATRIOT ROMAN      4

### 6d. How many copies of the film Hunchback Impossible exist in the inventory system?
There are 6 copies.

In [162]:
query = ("SELECT COUNT(*)\n"
        "FROM inventory i\n"
        "WHERE film_id = (SELECT f.film_id\n"
                        "FROM film f\n"
                        "WHERE f.title ilike 'Hunchback Impossible')\n")
df = pd.read_sql(query, con=conn)
df

count
0      6

### 6e. Using the tables payment and customer and the JOIN command, list the total paid by each customer. List the customers alphabetically by last name:

In [164]:
query = ("SELECT c.last_name, c.first_name, SUM(p.amount) as total_payment\n"
        "FROM customer c\n"
        "JOIN payment p on c.customer_id = p.customer_id\n"
        "GROUP BY c.customer_id, c.last_name, c.first_name\n"
        "ORDER BY c.last_name;\n")
df = pd.read_sql(query, con=conn)
df

last_name first_name  total_payment
0           ABNEY     RAFAEL          97.79
1            ADAM  NATHANIEL         133.72
2           ADAMS   KATHLEEN          92.73
3       ALEXANDER      DIANA         105.73
4          ALLARD     GORDON         160.68
5           ALLEN    SHIRLEY         126.69
6         ALVAREZ   CHARLENE         114.73
7        ANDERSON       LISA         106.76
8          ANDREW       JOSE          96.75
9         ANDREWS        IDA          76.77
10         AQUINO      OSCAR          99.80
11           ARCE      HARRY         157.65
12      ARCHULETA     JORDAN         132.70
13      ARMSTRONG    MELANIE          92.75
14         ARNOLD   BEATRICE         119.74
15      ARSENAULT       KENT         134.73
16          ARTIS       CARL         106.77
17       ASHCRAFT     DARRYL          76.77
18          ASHER     TYRONE         112.76
19         AUSTIN       ALMA         151.65
20         BAILEY    MILDRED          98.75
21          BAKER     PAMELA          95.77
22          BALES     MARTIN         103.73
23          BANDA    EVERETT         110.72
24          BANKS     JESSIE          91.74
25         BARBEE    CLAYTON          96.74
26        BARCLAY      ANGEL         115.68
27       BARFIELD   NICHOLAS         145.68
28        BARKLEY     VICTOR          91.76
29         BARNES     RACHEL          84.78
..            ...        ...            ...
569        WATSON    THERESA          99.70
570         WATTS     SHELLY         113.74
571         WAUGH      JAMIE         118.75
572           WAY       MIKE         166.65
573        WEAVER    YOLANDA         110.73
574          WEBB      ETHEL         135.68
575        WEINER     RONALD         132.70
576         WELCH    MARLENE         117.74
577         WELLS     SHEILA          73.82
578          WEST       EDNA         107.74
579  WESTMORELAND   MITCHELL         134.68
580         WHEAT       FRED          88.75
581       WHEELER       LUCY          91.74
582         WHITE      BETTY         117.72
583       WHITING        ROY         143.71
584         WILES        JON          87.76
585      WILLIAMS      LINDA         135.74
586    WILLIAMSON       GINA         111.72
587        WILLIS    BERNICE         145.67
588        WILSON      SUSAN          92.76
589       WINDHAM     DARREN         108.76
590       WOFFORD     VIRGIL         107.73
591          WOOD       LORI         141.69
592         WOODS   FLORENCE         126.70
593          WREN      TYLER          88.79
594        WRIGHT     BRENDA         104.74
595         WYMAN      BRIAN          52.88
596         YANEZ       LUIS          79.80
597           YEE     MARVIN          75.79
598         YOUNG    CYNTHIA         111.68

[599 rows x 3 columns]

### 7a. The music of Queen and Kris Kristofferson have seen an unlikely resurgence. As an unintended consequence, films starting with the letters K and Q have also soared in popularity. display the titles of movies starting with the letters K and Q whose language is English.


In [165]:
query = ("SELECT f.title\n"
        "FROM film f\n"
        "JOIN language l ON f.language_id = l.language_id AND name = 'English'\n"
        "WHERE f.title ILIKE 'k%' OR f.title ILIKE 'q%';\n")
df = pd.read_sql(query, con=conn)
df

title
0      KANE EXORCIST
1        KARATE MOON
2   KENTUCKIAN GIANT
3      KICK SAVANNAH
4   KILL BROTHERHOOD
5    KILLER INNOCENT
6     KING EVOLUTION
7         KISS GLORY
8      KISSING DOLLS
9      KNOCK WARLOCK
10  KRAMER CHOCOLATE
11     KWAI HOMEWARD
12        QUEEN LUKE
13   QUEST MUSSOLINI
14       QUILLS BULL

### 7b. Use subqueries to display all actors who appear in the film Alone Trip.


In [166]:
query = ("SELECT f.title, a.first_name, a.last_name\n"
        "FROM film_actor fa\n"
        "JOIN actor a ON fa.actor_id = a.actor_id\n"
        "JOIN film f ON fa.film_id = f.film_id\n"
        "WHERE title ILIKE \'ALone Trip\';\n")
df = pd.read_sql(query, con=conn)
df

title first_name last_name
0  ALONE TRIP         ED     CHASE
1  ALONE TRIP       KARL     BERRY
2  ALONE TRIP        UMA      WOOD
3  ALONE TRIP      WOODY     JOLIE
4  ALONE TRIP    SPENCER      DEPP
5  ALONE TRIP      CHRIS      DEPP
6  ALONE TRIP   LAURENCE   BULLOCK
7  ALONE TRIP      RENEE      BALL

### 7c. You want to run an email marketing campaign in Canada, for which you will need the names and email addresses of all Canadian customers. Use joins to retrieve this information.


In [189]:
query = ("SELECT c.first_name, c.last_name, ctry.country\n"
        "FROM customer c\n"
        "JOIN address a ON c.address_id = a.address_id\n"
        "JOIN city city ON a.city_id = city.city_id\n"
        "JOIN country ctry ON city.country_id = ctry.country_id\n"
        "WHERE ctry.country ILIKE \'Canada\';\n")
df = pd.read_sql(query, con=conn)
df

first_name  last_name country
0    DERRICK    BOURQUE  Canada
1    DARRELL      POWER  Canada
2    LORETTA  CARPENTER  Canada
3     CURTIS       IRBY  Canada
4       TROY    QUIGLEY  Canada

### 7d. Sales have been lagging among young families, and you wish to target all family movies for a promotion. Identify all movies categorized as a family film.  Now we mentioned family film, but there is no family film category. There’s a category that resembles that. In the real world nothing will be exact.


In [169]:
query = ("SELECT f.title\n"
        "FROM film f\n"
        "JOIN film_category fc on f.film_id = fc.film_id\n"
        "JOIN category c on fc.category_id = c.category_id AND name = 'Family';\n")
df = pd.read_sql(query, con=conn)
df

title
0              AFRICAN EGG
1            APACHE DIVINE
2           ATLANTIS CAUSE
3          BAKED CLEOPATRA
4                BANG KWAI
5        BEDAZZLED MARRIED
6          BILKO ANONYMOUS
7          BLANKET BEVERLY
8          BLOOD ARGONAUTS
9           BLUES INSTINCT
10        BRAVEHEART HUMAN
11           CHASING FIGHT
12         CHISUM BEHAVIOR
13          CHOCOLAT HARRY
14        CONFUSED CANDLES
15   CONVERSATION DOWNHILL
16              DATE SPEED
17      DINOSAUR SECRETARY
18              DUMBO LUST
19        EARRING INSTINCT
20        EFFECT GLADIATOR
21            FEUD FROGMEN
22        FINDING ANACONDA
23       GABLES METROPOLIS
24             GANDHI KWAI
25      GLADIATOR WESTWARD
26            GREASE YOUTH
27           HALF OUTFIELD
28             HOCUS FRIDA
29          HOMICIDE PEACH
..                     ...
39          MAGUIRE APACHE
40      MANCHURIAN CURTAIN
41       MOVIE SHAKESPEARE
42          MUSIC BOONDOCK
43           NATURAL STOCK
44            NETWORK PEAK
45             ODDS BOOGIE
46       OPPOSITE NECKLACE
47          PILOT HOOSIERS
48    PITTSBURGH HUNCHBACK
49          PRESIDENT BANG
50         PRIX UNDEFEATED
51              RAGE GAMES
52        RANGE MOONWALKER
53          REMEMBER DIARY
54  RESURRECTION SILVERADO
55          ROBBERY BRIGHT
56         RUSH GOODFELLAS
57        SECRETS PARADISE
58        SENSIBILITY REAR
59             SIEGE MADRE
60              SLUMS DUCK
61             SOUP WISDOM
62       SPARTACUS CHEAPER
63            SPINAL ROCKY
64             SPLASH GUMP
65            SUNSET RACER
66           SUPER WYOMING
67        VIRTUAL SPOILERS
68            WILLOW TRACY

[69 rows x 1 columns]

### 7e. Display the most frequently rented movies in descending order.


In [170]:
query = ("SELECT f.title, COUNT(*) as rental_cnt\n"
        "FROM rental r\n"
        "JOIN inventory i ON r.inventory_id = i.inventory_id\n"
        "JOIN film f ON i.film_id = f.film_id\n"
        "GROUP BY f.film_id, f.title\n"
        "ORDER BY rental_cnt DESC;\n")
df = pd.read_sql(query, con=conn)
df

title  rental_cnt
0         BUCKET BROTHERHOOD          34
1           ROCKETEER MOTHER          33
2             JUGGLER HARDLY          32
3              SCALAWAG DUCK          32
4             FORWARD TEMPLE          32
5             GRIT CLOCKWORK          32
6        RIDGEMONT SUBMARINE          32
7               ROBBERS JOON          31
8             TIMBERLAND SKY          31
9                  WIFE TURN          31
10             APACHE DIVINE          31
11              NETWORK PEAK          31
12           RUSH GOODFELLAS          31
13              HOBBIT ALIEN          31
14         GOODFELLAS SALUTE          31
15                 ZORRO ARK          31
16                FROST HEAD          30
17           IDOLS SNATCHERS          30
18              PULP BEVERLY          30
19           SUSPECTS QUILLS          30
20       RUGRATS SHAKESPEARE          30
21          ENGLISH BULWORTH          30
22               SHOCK CABIN          30
23              DOGMA FAMILY          30
24        BUTTERFLY CHOCOLAT          30
25             CAT CONEHEADS          30
26             GRAFFITI LOVE          30
27             WITCHES PANIC          30
28               HARRY IDAHO          30
29            MASSACRE USUAL          30
..                       ...         ...
928          ITALIAN AFRICAN           6
929         RUSHMORE MERMAID           6
930            SCHOOL JACKET           6
931             BED HIGHBALL           6
932  DESPERATE TRAINSPOTTING           6
933          TERMINATOR CLUB           6
934               SLING LUKE           6
935         OKLAHOMA JUMANJI           6
936              SIMON NORTH           6
937       GRACELAND DYNAMITE           6
938         WARLOCK WEREWOLF           6
939     APOCALYPSE FLAMINGOS           6
940               DUCK RACER           6
941        CONSPIRACY SPIRIT           5
942          MANNEQUIN WORST           5
943             HUNTER ALTER           5
944         BRAVEHEART HUMAN           5
945             FEVER EMPIRE           5
946           TRAFFIC HOBBIT           5
947          FULL FLATLINERS           5
948              BUNCH MINDS           5
949          INFORMER DOUBLE           5
950              GLORY TRACY           5
951        FREEDOM CLEOPATRA           5
952             PRIVATE DROP           5
953              SEVEN SWARM           5
954       MUSSOLINI SPOILERS           5
955              TRAIN BUNCH           4
956              MIXED DOORS           4
957           HARDLY ROBBERS           4

[958 rows x 2 columns]

### 7f. Write a query to display how much business, in dollars, each store brought in.


In [171]:
query = ("SELECT s.store_id, CAST(SUM(p.amount) AS money) AS store_revenue\n"
        "FROM rental r\n"
        "JOIN payment p ON r.rental_id = p.rental_id\n"
        "JOIN staff s ON r.staff_id = s.staff_id\n"
        "GROUP BY s.store_id;\n")
df = pd.read_sql(query, con=conn)
df

store_id store_revenue
0         1    $33,534.57
1         2    $33,881.94

### 7g. Write a query to display for each store its store ID, city, and country.

In [172]:
query = ("SELECT s.store_id, c.city, ctry.country\n"
        "FROM store s\n"
        "JOIN address a ON a.address_id = s.address_id\n"
        "JOIN city c on a.city_id = c.city_id\n"
        "JOIN country ctry on c.country_id = ctry.country_id;\n")
df = pd.read_sql(query, con=conn)
df

store_id        city    country
0         1  Lethbridge     Canada
1         2   Woodridge  Australia

### 7h. List the top five genres in gross revenue in descending order. 

In [173]:
query = ("SELECT c.name as genre, CAST(SUM(p.amount) AS money) as genre_revenue\n"
        "FROM payment p\n"
        "JOIN rental r ON p.rental_id = r.rental_id\n"
        "JOIN inventory i ON r.inventory_id = i.inventory_id\n"
        "JOIN film_category fc on i.film_id = i.film_id\n"
        "JOIN category c on fc.category_id = c.category_id\n"
        "GROUP BY c.category_id, c.name\n"
        "ORDER BY genre_revenue DESC\n"
        "LIMIT 5;\n")
df = pd.read_sql(query, con=conn)
df

genre  genre_revenue
0       Sports  $4,988,821.74
1      Foreign  $4,921,405.23
2       Family  $4,651,739.19
3  Documentary  $4,584,322.68
4    Animation  $4,449,489.66

### 8a. In your new role as an executive, you would like to have an easy way of viewing the Top five genres by gross revenue. Use the solution from the problem above to create a view. 


In [177]:
query1 = ("DROP VIEW IF EXISTS genre_revenue;")
query2 = ("CREATE VIEW genre_revenue AS\n"
    "SELECT c.name as genre, CAST(SUM(p.amount) AS money) as genre_revenue\n"
    "FROM payment p\n"
    "JOIN rental r ON p.rental_id = r.rental_id\n"
    "JOIN inventory i ON r.inventory_id = i.inventory_id\n"
    "JOIN film_category fc on i.film_id = i.film_id\n"
    "JOIN category c on fc.category_id = c.category_id\n"
    "GROUP BY c.category_id, c.name\n"
    "ORDER BY genre_revenue DESC\n"
    "LIMIT 5;\n")

cur = conn.cursor()
cur.execute(query1)
cur.execute(query2)


### 8b. How would you display the view that you created in 8a?

In [178]:
df = pd.read_sql('SELECT * FROM genre_revenue', con=conn)
df

genre  genre_revenue
0       Sports  $4,988,821.74
1      Foreign  $4,921,405.23
2       Family  $4,651,739.19
3  Documentary  $4,584,322.68
4    Animation  $4,449,489.66

### 8c. You find that you no longer need the view top_five_genres. Write a query to delete it.

In [179]:
query1 = ("DROP VIEW IF EXISTS genre_revenue;")
cur = conn.cursor()
cur.execute(query1)